### PR TITLE
Implement repr(packed) for repr(simd)

### PR DESCRIPTION
**Pull request by calebzulawski**
This allows creating vectors with non-power-of-2 lengths that do not have padding.  See rust-lang/portable-simd#319

### CODE REVIEW

File: compiler/rustc_ty_utils/src/layout.rs [hunk 1]
```diff
@@ -431,7 +431,21 @@ fn layout_of_uncached<'tcx>(
                 .size
                 .checked_mul(e_len, dl)
                 .ok_or_else(|| error(cx, LayoutError::SizeOverflow(ty)))?;
-            let align = dl.vector_align(size);
+
+            let (abi, align) = if def.repr().packed() && !e_len.is_power_of_two() {
+                // Non-power-of-two vectors have padding up to the next power-of-two.
```
Comment (programmerjake) [inline compiler/rustc_ty_utils/src/layout.rs:436]: iirc LLVM doesn't guarantee vectors don't have more padding than just to the next power of two, e.g. padding `<2 x i8>` to 128-bits.

**Comment by calebzulawski** [comment]
> Is it possible it would make sense for this just to be what `repr(simd)` _means_, without needing `packed`?
> 
> Since all the other uses of it are power-of-two anyway, and thus if you make `simd` always be the GCD, that might be fine?

Maybe eventually, but for now I didn't implement any way to call simd intrinsics on packed repr vectors--they still need to be loaded into full (padded) vectors.  std::simd will need to convert the vectors before calling intrinsics.

**Comment by calebzulawski** [comment]
Note that I also change the ABI, so it's not possible to pass these in vector registers (I think).  So if target feature calling conventions ever get worked out, there may still be a benefit to the current implementation as well.

File: compiler/rustc_ty_utils/src/layout.rs (hunk 1 above)
Comment (Sp00ph) [reply compiler/rustc_ty_utils/src/layout.rs:436]: https://llvm.org/docs/LangRef.html#vector-type seems to say that LLVM guarantees that vectors have no padding in their in memory representation? Padding `<2 x i8>` to 128 bits is something that LLVM does inside the vector registers (e.g. on aarch64), but never in memory.

**Comment by workingjubilee** [comment]
> Note that I also change the ABI, so it's not possible to pass these in vector registers (I think). So if target feature calling conventions ever get worked out, there may still be a benefit to the current implementation as well.

Given that we're specifically working on defining target feature minima specifically so we can do that, I don't think that blocking passing in registers is acceptable.

File: compiler/rustc_ty_utils/src/layout.rs (hunk 1 above)
Comment (programmerjake) [reply compiler/rustc_ty_utils/src/layout.rs:436]: [`core::mem::size_of`](https://doc.rust-lang.org/core/mem/fn.size_of.html):
> More specifically, this is the offset in bytes between successive elements in an array with that item type including alignment padding.

**Comment by calebzulawski** [comment]
Unfortunately I think we simply can't have it both ways. LLVM defines vectors as having padding and we would like them to not have padding. This change doesn't affect all vectors, just those with repr(packed), which specifically removes padding and reduces alignment.  A user could always use a non repr(packed) vector to pass by register.

**Comment by programmerjake** [comment]
> > Note that I also change the ABI, so it's not possible to pass these in vector registers (I think). So if target feature calling conventions ever get worked out, there may still be a benefit to the current implementation as well.
> 
> Given that we're specifically working on defining target feature minima specifically so we can do that, I don't think that blocking passing in registers is acceptable.

I don't expect this to block passing in registers, because this is defining the in-memory ABI. the passing-by-value ABI can be totally different, since semantically you have to then store that value somewhere in memory before you can look at the bytes.

**Comment by programmerjake** [comment]
> Unfortunately I think we simply can't have it both ways.

I think we can have it both ways -- just pass-by-value as a LLVM vector and use the array type for in-memory address calculations and allocating memory. load/store with vector types only read-write the non-padding bytes, so we're fine.

**Comment by calebzulawski** [comment]
> > Unfortunately I think we simply can't have it both ways.
> 
> I think we can have it both ways -- just pass-by-value as a LLVM vector and use the array type for in-memory address calculations and allocating memory. load/store with vector types only read-write the non-padding bytes, so we're fine.

True, this is definitely possible once we are able to.

File: compiler/rustc_ty_utils/src/layout.rs (hunk 1 above)
Comment (calebzulawski) [reply compiler/rustc_ty_utils/src/layout.rs:436]: Whoops, I meant rustc assumes that npot vectors round up to the next power of two.

Comment (programmerjake) [reply compiler/rustc_ty_utils/src/layout.rs:436]: > > that's only for load/store, LLVM vector types still have big enough alignment that Rust requires padding in many cases.
> 
> I thought LLVM types didn't have any sort of intrinsic alignment? And even if they have a preferred alignment that's too big, can't you just always annotate load/store instructions with the smaller alignment?

they do have an intrinsic ABI alignment, it's used for default alignment of load/store/alloca when not explicitly specified and for stack spilling and a few other things

**Comment by workingjubilee** [comment]
I don't think that will optimize well at all. LLVM historically reasons about in-memory and in-register data in conflationary ways.

**Comment by calebzulawski** [comment]
I don't think there's any optimization involved. It's just a single vector load.

Either way, this PR doesn't change how existing vectors work.  `repr(packed)` is specified to remove padding and it doesn't work correctly on SIMD types, which this change fixes.

I agree that the calling convention could become an issue in the future, but I'd like to deal with that in the future when it comes up, since there's still no consensus on a fix.  Also, fixing `repr(packed)` for SIMD doesn't require us to use it for std::simd, but I'd like to fix this issue first so we can try it out.

**Comment by workingjubilee** [comment]
"Legalizing to a vector load instead of a series of scalar loads" is an optimization.

**Comment by workingjubilee** [comment]
...Okay, I didn't quite divine from the messages/code changes that `repr(packed)` doesn't have _any_ functionality currently with `repr(simd)`, and that we don't even emit an error.

**Comment by workingjubilee** [comment]
I would prefer this to come with codegen/assembly tests to demonstrate what it actually looks like when these types are interacted with, and to prove it like... legalizes correctly, where "correctly" in this case is "not an LLVM error, I guess?"

**Comment by calebzulawski** [comment]
Since `repr(simd, packed)` can't really be used directly, I added a pretty simple test that just loads a packed vector and performs an operation on it, just to make sure it doesn't crash LLVM or anything like that.

File: compiler/rustc_ty_utils/src/layout.rs (hunk 1 above)
Comment (programmerjake) [reply compiler/rustc_ty_utils/src/layout.rs:436]: > I'm not sure if it's guaranteed, but rustc does make the assumption (somewhere, I forget where)

just found this:
https://lang-team.rust-lang.org/frequently-requested-changes.html#size--stride
> Rust assumes that the size of an object is equivalent to the stride of an object

**Comment by workingjubilee** [comment]
I want to approve this but I still can't see what the emitted LLVMIR is.

**Comment by calebzulawski** [comment]
Not quite as rigorous as as a codegen test, consider this reduced case for illustrative purposes:
```rust
#![feature(repr_simd, platform_intrinsics)]

#[repr(simd, packed)]
pub struct Simd<T, const N: usize>([T; N]);

#[repr(simd)]
#[derive(Copy, Clone)]
pub struct FullSimd<T, const N: usize>([T; N]);

extern "platform-intrinsic" {
    fn simd_mul<T>(a: T, b: T) -> T;
}

// non-powers-of-two have padding and need to be expanded to full vectors
fn load<T, const N: usize>(v: Simd<T, N>) -> FullSimd<T, N> {
    unsafe {
        let mut tmp = core::mem::MaybeUninit::<FullSimd<T, N>>::uninit();
        std::ptr::copy_nonoverlapping(&v as *const _, tmp.as_mut_ptr().cast(), 1);
        tmp.assume_init()
    }
}

pub fn square(x: Simd<f32, 3>) -> FullSimd<f32, 3> {
    let x = load(x);
    unsafe { simd_mul(x, x) }
}
  ```

With optimization, this simply generates the following:
```
; foo::square
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) uwtable
define void @_ZN3foo6square17h0c2320cc22379be1E(ptr noalias nocapture noundef writeonly sret(<3 x float>) align 16 dereferenceable(16) %_0, ptr noalias nocapture noundef readonly align 4 dereferenceable(12) %x) unnamed_addr #0 {
start:
  %x.val = load <3 x float>, ptr %x, align 4
  %0 = fmul <3 x float> %x.val, %x.val
  store <3 x float> %0, ptr %_0, align 16
  ret void
}
```
This is nearly the same as using regular vectors, but note that the input `%x` is passed as a pointer with `align 4`.  Remember that vectors are always passed by reference anyway, but it would be `align 16`.  You will also see on the first line, where a regular vector load is `align 16

**Comment by calebzulawski** [comment]
Whoops, accidentally closed while writing my comment :)

**Comment by calebzulawski** [comment]
@workingjubilee I added my comment above as a codegen test

**Comment by calebzulawski** [comment]
I'm a little lost, those tests pass on my machine.  It appears it's not optimizing here.

**Comment by calebzulawski** [comment]
@workingjubilee I think this is probably good enough for a codegen test? It doesn't rely on any optimization.

**Comment by workingjubilee** [comment]
Yep! Looks good enough for me atm,
@bors r+

**Comment by calebzulawski** [comment]
@workingjubilee I removed the codegen test since it's so dependent on optimization (just little things like attributes are changing, but making it impossible to write), and the tests run on so many optimization levels, I can't seem to make a useful test.  Considering `repr(simd)` is unstable in the first place, hopefully you've seen enough codegen examples to be confident enough to merge as-is

**Comment by workingjubilee** [comment]
Indeed, I mostly wanted an example! sadness about the test, though. It really shouldn't be so hard...

@bors r+

**Comment by calebzulawski** [comment]
Thanks :)

File: tests/ui/simd/repr_packed.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+// run-pass
+
+#![feature(repr_simd, platform_intrinsics)]
+#![allow(non_camel_case_types)]
+
+#[repr(simd, packed)]
+struct Simd<T, const N: usize>([T; N]);
```
Comment (RalfJung) [inline tests/ui/simd/repr_packed.rs:7]: From how `packed` usually works, I would expect this to mean that the type has alignment 1. But that doesn't seem to be the case, instead the alignment is the largest possible for the size, or something like that?

What happens with `packed(N)`?

Would be good to have the interaction of `simd` and `packed` documented somewhere.

**Comment by RalfJung** [comment]
FWIW codegen has support for using different LLVM types in by-val vs by-ref situations: specifically, `bool` is `i8` in memory but `i1` as an SSA local. Maybe something similar could be done for these packed simd types?

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 3]
```diff
@@ -946,6 +946,13 @@ fn generic_simd_intrinsic<'ll, 'tcx>(
         tcx.normalize_erasing_late_bound_regions(ty::ParamEnv::reveal_all(), callee_ty.fn_sig(tcx));
     let arg_tys = sig.inputs();
 
+    // Vectors must be immediates (non-power-of-2 #[repr(packed)] are not)
+    for (ty, arg) in arg_tys.iter().zip(args) {
+        if ty.is_simd() && !matches!(arg.val, OperandValue::Immediate(_)) {
+            return_error!(InvalidMonomorphization::SimdArgument { span, name, ty: *ty });
+        }
+    }
```
Comment (RalfJung) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:954]: This comment seems no longer accurate... maybe that got changed by https://github.com/rust-lang/rust/pull/125311? Doing simd_mul etc on a packed SIMD type works just fine now. Even in debug builds this [generates the code one would hope for](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=49bba1357ef904726ba123e3afbc3545).

**Pull request merged**